### DISCUSSION TITLE
Understanding Maven Repository Resolution

### DISCUSSION THREAD

**Discussion by kjq** [renovatebot/renovate#29903] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

GitLab

### Please tell us more about your question or problem

I am trying to better understand how Renovate is resolving Maven repositories.

We have a private repository which houses our dependencies and then rely on Maven Central for the rest.  In our Maven pom.xml they are pretty plain because the settings.xml is store per-user (`~/.m2/setttings.xml`) not with each project.  Because of this we do not need to define a `<repository/>` element in the pom.xml.  When running locally, we can see the dependencies being pulled from the right locations (mostly Maven Central) and ours coming from our repository.

No matter what I do in my Renovate config I seem to only achieve an either/or results - it either goes to Maven Central for everything which fails on our internal dependencies or it goes to a Maven group on our side which retrieves everything from us.

I've tried a `packageRule` like this thinking first it would hit central, then if not found hit our repo.

```json
{
    "matchCategories": [
        "java"
    ],
    "registryUrls": [
        "https://repo1.maven.org/maven2/",
        "https://my-repo/big-group"
    ]
},
```

Without having the logs in front of me it does not even seem to be merging them in the config and it looks something like this:

```json
{
    "datasource": "maven",
    "depName": "my.company.artifact",
    "currentValue": "${revision}${changelist}",
    "fileReplacePosition": 1182,
    "registryUrls": ["https://repo.maven.apache.org/maven2"],
    "depType": "parent-root",
    "skipReason": "version-placeholder",
    "updates": [],
    "packageName": "my.company:artifact"
},
```

So, essentially, how i can configure two repositories, one private and the other Maven Central, so that if an artifact does not exist in one it uses the other?

**Comment by Churro** [comment]
Probably try a different matcher because as stated [here](https://docs.renovatebot.com/configuration-options/#matchcategories):
> Rules with matchCategories are only applied after extraction of dependencies. If you want to configure which managers are being extracted at all, use enabledManagers instead.

**Reply by kjq** [reply]
That makes sense.  I didnt catch that part about the "after" extraction.  Can I use different "matchCriteria"?  It looks like if i use `enableManagers` i will have to turn on each one i am using.

**Reply by Churro** [reply]
Using `"matchManagers": ["maven"]` could work in your case.

**Reply by kjq** [reply]
Its pretty hard to tell but the logs show what looks to be it is hitting both repositories now equally (and the requests being made seem to match).  I would expect the "my.repo" to only have a handful for dependencies if found in maven central first (which the logs are search first too then my repo).

```json
        {
            "matchManagers": [
                "maven"
            ],
            "registryUrls": [
                "https://repo1.maven.org/maven2/",
                "https://my.repo"
            ]
        },
```

```json
         "my.repo": {
           "count": 98,
           "reqAvgMs": 103,
           "reqMedianMs": 74,
           "reqMaxMs": 379,
           "queueAvgMs": 0,
           "queueMedianMs": 0,
           "queueMaxMs": 1
         },

        "repo1.maven.org": {
           "count": 98,
           "reqAvgMs": 65,
           "reqMedianMs": 51,
           "reqMaxMs": 340,
           "queueAvgMs": 0,
           "queueMedianMs": 0,
           "queueMaxMs": 3
         }
```

**Reply by Churro** [reply]
Packages are sometimes moved between registries, e.g., from jcenter to maven central, and querying only one might be incomplete. Also the order of registries is not taken into account with all managers. If you want to see more targeted queries, you can refine your package rule to match your internal packages e.g., via `matchPackagePatterns` and then specify only the internal registry URL for those.

**Reply by kjq** [reply]
That may be fine but looking at the logs it looks to query both repositories equally.  Also, I would of thought if it found it in one it stopped and would not try the other repository.

The reason we are trying to minimize the repository calls is because even running Renovate for a short period of time during off hours is putting a lot of unneeded pressure on the Nexus.

Another option we are thinking about is spacing out our renovate jobs so they are less frequent.

I thought about setting the packages as well but that will be a nightmare for so many different projects but may be the last resort.

One other option I wanted to try but maybe you can tell me if it would be similar or not, is to use our settings.xml that is on the build runner.  The question there is how to use that for all projects because it is not committed with each.

**Reply by Churro** [reply]
Using `settings.xml` should work, see https://docs.renovatebot.com/java/#maven-file-support

**Reply by kjq** [reply]
I saw that but the settings.xml is not checked in with each projects.  It is place at a location on the CI runner and used across any Maven jobs.  How can i configure it to use a location rather than per-project settings?

**Reply by Churro** [reply]
Afaik that's not supported

**Reply by kjq** [reply]
Is there a pre-step I can do when renovate clones the project to "copy" the settings.xml into that project or match it somehow?  The issue is that the build server has a much different settings with "push" access where as day to day usage locally is read only.  I might try a settings.xml just to see what happens though.

**Reply by Churro** [reply]
There is no way in renovate to extend the repo workspace after cloning. If the changes you need affect multiple repos, you could centralize them in a preset file that is stored outside of the repo but referenced from within your project's local `renovate.json` -> https://docs.renovatebot.com/key-concepts/presets/

**Reply by kjq** [reply]
As a note, i just tried creating a `.mv/settings.xml` in one project and then running Renovate against it.  It looked to have no effect.

I also removed the packageRule we have been discussing assuming it would not be needed anymore.

```
DEBUG: Found registryUrls in settings.xml (repository=myproject/artifact, packageFile=.mvn/settings.xml)
```

The dependencies get listed like and it does not even go to my repo only central.

```
{
                 "datasource": "maven",
                 "depName": "xxxxxxxxxxxxxxxxxxxxx",
                 "currentValue": "${revision}${changelist}",
                 "fileReplacePosition": 708,
                 "registryUrls": ["https://repo.maven.apache.org/maven2"],
                 "depType": "parent-root",
                 "skipReason": "version-placeholder",
                 "updates": [],
                 "packageName": "xxxxxxxxxxxxxxxxxxxx"
               },
```

**Reply by Churro** [reply]
As can be seen in your log snippet, renovate skipped it since it found two variables _revision_ and _changelist_. As they are not resolved at this point, the item is skipped.

**Reply by kjq** [reply]
So since it was skipped at that point the `registryUrls` would not of been updated?  As a note, we use Maven "friendly" (or not so friendly depending on how you look at it) versioning and that has been working so far.

**Reply by kjq** [reply]
Btw, thank you for responding and helping me work through this.  It is very appreciated.

**Reply by Churro** [reply]
If I'm not mistaken the registry URL package rule should be applied to extracted deps even if it was then skipped. In the debug log you should also find a line like `Found registryUrls in settings.xml`.

**Reply by kjq** [reply]
I did see this in the logs but thought the dependency would of then reflected it:

```
DEBUG: Found registryUrls in settings.xml (repository=zzzzzzzzzzzzz, packageFile=.mvn/settings.xml)
       "registries": []
DEBUG: manager extract durations (ms) (repository=zzzzzzzzzzzzzzzzzzz)
       "managers": {
         "gitlabci": 1,
         "gitlabci-include": 1,
         "kubernetes": 11,
         "dockerfile": 70,
         "maven": 113
       }
```

So maybe that works when the settings is in the XML?  When I dont have it the dependencies look the same but i see it making the calls to maven central and our repository.

**Reply by Churro** [reply, marked as answer]
I'm not entirely sure what you're up to now: if you add registries into a `settings.xml` and put it into your repository, the list of extracted repos will be added to the overall list. The call to Maven central will still happen, since this is the general fallback registry in maven. 

You can only disable this fallback to Maven central by overwriting registry URLs explicitly using a matching pattern. As can be seen in renovate docs, there are different matching patterns. `matchPackagePatterns` is only one option. You could also set `"matchManagers": ["maven"]` or related options + the list of desired registry URLs. This way renovate won't visit Maven Central at all. 

The maven datasource in renovate implements a `merge` strategy, i.e., visiting all specified repos and then merging version results. This was deliberately implemented to avoid situations like https://github.com/renovatebot/renovate/discussions/25461.

**Reply by kjq** [reply]
Im going to step back for a bit and look more into this over the weekend and update you.  I've been through so many iterations at this point I cant remember what did what anymore...
